### PR TITLE
Use `set -o exportall` to write less code

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
# Description

Addresses https://github.com/kerl/kerl/issues/194#issuecomment-379368047.

**NOTE**: not to be merged yet, since this builds on top of https://github.com/kerl/kerl/pull/454. I'll rebase once the other one gets approved+merged.

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/kerl/kerl/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: kerl [hunk 1]
```diff
@@ -1197,90 +1198,49 @@ do_build_install() {
...
-    [ -n "$KERL_RELEASE_TARGET" ] && [ -f bin/cerl ] && cp bin/cerl "$absdir/bin"
-
-    list_add installations "$1 $absdir";
-    cat <<ACTIVATE >"$absdir"/activate
+  cat <<\
+=======
 #!/bin/sh
 # credits to virtualenv
 kerl_deactivate() {
+    set -o allexport
```
Comment (paulo-ferraz-oliveira) [inline kerl:1215]: Relevant bit 1 starts here... (look at all those `export` go away)

File: kerl [hunk 2]
```diff
@@ -1291,50 +1251,67 @@ kerl_deactivate() {
     fi
     unset KERL_ENABLE_PROMPT
     unset KERL_PROMPT_FORMAT
+    set +o allexport
```
Comment (paulo-ferraz-oliveira) [inline kerl:1254]: Relevant bit 1 ends here...

File: kerl [hunk 3]
```diff
@@ -1291,50 +1251,67 @@ kerl_deactivate() {
...
+if [ -n "\$_KERL" ]; then
+  _KERL_VERSION=\$(\$_KERL version)
+fi
+
+if [ -n "\$_KERL_VERSION" ] && [ "\$_KERL_VERSION" != "$KERL_VERSION" ]; then
+  echo "WARNING: this Erlang/OTP installation appears to be stale. Please consider reinstalling."
+  echo "         It was created with kerl $KERL_VERSION, and the current version is \$_KERL_VERSION."
+fi
+
+set -o allexport
```
Comment (paulo-ferraz-oliveira) [inline kerl:1269]: Relevant bit 2 starts here... (look at all those `export` go away)

File: kerl [hunk 4]
```diff
@@ -1291,50 +1251,67 @@ kerl_deactivate() {
...
-    PS1="\$PRMPT\$PS1"
-    export PS1
+    _KERL_PRMPT=\$(echo "\$_KERL_PROMPT_FORMAT" | sed 's^%RELEASE%^$release^;s^%BUILDNAME%^$build_name^')
+    PS1="\$_KERL_PRMPT\$PS1"
 fi
 if [ -n "\$BASH" ] || [ -n "\$ZSH_VERSION" ]; then
     hash -r
 fi
-ACTIVATE
+set +o allexport
```
Comment (paulo-ferraz-oliveira) [inline kerl:1300]: Relevant bit 2 ends here...

File: kerl [hunk 5]
```diff
@@ -1291,50 +1251,67 @@ kerl_deactivate() {
...
-    export ERL_AFLAGS="-kernel shell_history enabled \$ERL_AFLAGS"
+    ERL_AFLAGS="-kernel shell_history enabled \$ERL_AFLAGS"
 fi
 # shellcheck source=/dev/null
 if [ -f "$KERL_CONFIG" ]; then . "$KERL_CONFIG"; fi
 if [ -n "\$KERL_ENABLE_PROMPT" ]; then
     _KERL_SAVED_PS1="\$PS1"
-    export _KERL_SAVED_PS1
     if [ -n "\$KERL_PROMPT_FORMAT" ]; then
-        FRMT="\$KERL_PROMPT_FORMAT"
```
Comment (paulo-ferraz-oliveira) [inline kerl:1324]: Renamed in order to be "safely" exported (though this is a local-only variable).

File: kerl [hunk 6]
```diff
@@ -1291,50 +1251,67 @@ kerl_deactivate() {
...
+        _KERL_PROMPT_FORMAT="\$KERL_PROMPT_FORMAT"
     else
-        FRMT="(%BUILDNAME%)"
+        _KERL_PROMPT_FORMAT="(%BUILDNAME%)"
     fi
-    PRMPT=\$(echo "\$FRMT" | sed 's^%RELEASE%^$rel^;s^%BUILDNAME%^$1^')
```
Comment (paulo-ferraz-oliveira) [inline kerl:1328]: Renamed in order to be "safely" exported (though this is a local-only variable).

File: kerl [hunk 7]
```diff
@@ -1256,47 +1251,53 @@ kerl_deactivate() {
...
+    ERL_AFLAGS="-kernel shell_history enabled \$ERL_AFLAGS"
 fi
 # shellcheck source=/dev/null
 if [ -f "$KERL_CONFIG" ]; then . "$KERL_CONFIG"; fi
 if [ -n "\$KERL_ENABLE_PROMPT" ]; then
     _KERL_SAVED_PS1="\$PS1"
-    export _KERL_SAVED_PS1
     if [ -n "\$KERL_PROMPT_FORMAT" ]; then
-        FRMT="\$KERL_PROMPT_FORMAT"
+        _KERL_PROMPT_FORMAT="\$KERL_PROMPT_FORMAT"
```
Comment (paulo-ferraz-oliveira) [inline kerl:1290]: Renamed in order to be "safely" exported (though this is a local-only variable).

File: kerl [hunk 8]
```diff
@@ -1256,47 +1251,53 @@ kerl_deactivate() {
...
     else
-        FRMT="(%BUILDNAME%)"
+        _KERL_PROMPT_FORMAT="(%BUILDNAME%)"
     fi
-    PRMPT=\$(echo "\$FRMT" | sed 's^%RELEASE%^$release^;s^%BUILDNAME%^$build_name^')
-    PS1="\$PRMPT\$PS1"
-    export PS1
+    _KERL_PRMPT=\$(echo "\$_KERL_PROMPT_FORMAT" | sed 's^%RELEASE%^$release^;s^%BUILDNAME%^$build_name^')
```
Comment (paulo-ferraz-oliveira) [inline kerl:1294]: Renamed in order to be "safely" exported (though this is a local-only variable).

File: kerl [hunk 9]
```diff
@@ -1212,40 +1212,35 @@ emit_activate() {
 #!/bin/sh
 # credits to virtualenv
 kerl_deactivate() {
+    set -o allexport
```
Comment (jadeallenx) [inline kerl:1215]: does this bleed over into the login shell or does this only affect kerl?

Comment (paulo-ferraz-oliveira) [reply kerl:1215]: There's a `set +o allexport` at the end, so unless this breaks in the middle of the execution and we're doing some other concurrent thing (not sure if this is even possible) I'd say it's not relevant. Do you prefer me to add a test? But what kind of test would that be?

**Review by jadeallenx — APPROVED** [review]

**Pull request merged**